### ISSUE TITLE
RestKnowledgeBase::toBeProcessedByKnowledgeBase should be thread-safe

### ISSUE THREAD

**Issue by Sophietje** [TNO/knowledge-engine#375]
In GitLab by @han.kruiger.tno.nl on Mar 28, 2022, 16:14

Adding `synchronized` to it seems to solve an issue, but a less bulky solution is preferred.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Mar 29, 2022, 06:31

Maybe the solution below would be enough:

<pre>
boolean success = false;
if (asyncResponse != null) {
  //stop longpoll and send handle request to client

  <b>success = asyncResponse.resume(...);</b>
}

if (asyncResponse == null && <b>!success</b>)
{
  //add handle request to queue
}
</pre>

But maybe there are more elegant ways and this solution depends on whether the asyncResponse.resume() method actually returns `false` if the second thread arrives there.

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Mar 29, 2022, 07:22

created branch [`361-restknowledgebase-tobeprocessedbyknowledgebase-is-not-thread-safe`](/interconnect/knowledge-engine/-/compare/master...361-restknowledgebase-tobeprocessedbyknowledgebase-is-not-thread-safe) to address this issue

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Mar 29, 2022, 08:43

I improved it with a solution similar to yours, but I reasoned that we'd still need a lock! Because if the method is called simultaneously from two threads, the `asyncResponse` object will be overwritten to `null` by one thread, while the other one still thinks it's a valid object. (In fact, the *reference* to the object might not even have been updated and the `asyncResponse.resume` call would not succeed because it already resumed.)

Anyway, it should work better now! :smile:
